pkg/cmd/server/crypto/crypto.go

```diff
@@ -142,15 +143,82 @@ var (
 )
 
 type CA struct {
+	Config *TLSCertificateConfig
+
+	SerialGenerator SerialGenerator
+}
+
+// SerialGenerator is an interface for getting a serial number for the cert.  It MUST be thread-safe.
+type SerialGenerator interface {
+	Next(template *x509.Certificate) (int64, error)
+}
+
+// SerialFileGenerator returns a unique, monotonically increasing serial number and ensures the CA on disk records that value.
+type SerialFileGenerator struct {
 	SerialFile string
-	Config     *TLSCertificateConfig
 
 	// lock guards access to the Serial field
 	lock   sync.Mutex
 	Serial int64
 }
 
+func NewSerialFileGenerator(serialFile string, createIfNeeded bool) (*SerialFileGenerator, error) {
+	// read serial file
+	var serial int64
+	serialData, err := ioutil.ReadFile(serialFile)
+	if err == nil {
+		serial, _ = strconv.ParseInt(string(serialData), 16, 64)
+	}
+	if os.IsNotExist(err) && createIfNeeded {
+		if err := ioutil.WriteFile(serialFile, []byte("0"), 0644); err != nil {
+			return nil, err
+		}
+		serial = 1
+	}
+
+	if serial < 1 {
+		serial = 1
+	}
+
+	return &SerialFileGenerator{
+		Serial:     serial,
+		SerialFile: serialFile,
+	}, nil
+}
+
+// Next returns a unique, monotonically increasing serial number and ensures the CA on disk records that value.
+func (s *SerialFileGenerator) Next(template *x509.Certificate) (int64, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	next := s.Serial + 1
+	s.Serial = next
+
+	// Output in hex, padded to multiples of two characters for OpenSSL's sake
+	serialText := fmt.Sprintf("%X", next)
+	if len(serialText)%2 == 1 {
+		serialText = "0" + serialText
+	}
+
+	if err := ioutil.WriteFile(s.SerialFile, []byte(serialText), os.FileMode(0640)); err != nil {
+		return 0, err
+	}
+	return next, nil
+}
+
+// HashingSerialGenerator returns a serial based on time.Now and the subject
+type HashingSerialGenerator struct {
+}
+
+func (s *HashingSerialGenerator) Next(template *x509.Certificate) (int64, error) {
+	str := fmt.Sprintf("%v %v", template.Subject, time.Now())
```

Comment:
> I would actually just go full random, rather than a predictable timestamp

Made random, but it is predictable.  I'm not sure if it matters for this usage though.